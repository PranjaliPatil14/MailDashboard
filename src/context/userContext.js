import React, { useEffect, useState } from "react";
import * as PropTypes from "prop-types";
import useLocalStorage, { USER } from "../utilities/useLocalStorage";

const initialUserContextValue = {
  users: [],
  loggedInUser: {
    id: 0,
    name: "",
    role: "",
    profileImage: "",
  },
  setLoggedInUser: () => {},
  setUsers: () => {},
};

export const UserContext = React.createContext(initialUserContextValue);

const UserContextProvider = ({ children }) => {
  const [users, setUsers] = useState([]);
  const [loggedInUser, setLoggedInUser] = useState({});
  const { value } = useLocalStorage(USER);

  useEffect(() => {
    if (!users.length) {
      setUsers(value);
    }
  }, [users.length, value]);

  return (
    <UserContext.Provider
      value={{ users, setUsers, loggedInUser, setLoggedInUser }}
    >
      {children}
    </UserContext.Provider>
  );
};

UserContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default UserContextProvider;
