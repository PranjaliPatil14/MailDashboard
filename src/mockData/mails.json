[
  {"id": 1, "from":1,"to":[3],"cc":[2],"body":"Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla eget rhoncus metus.","attachment":false,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"Development", "read": false, "important": false},
  {"id": 2, "from":2,"to":[1],"cc":[3],"body":"Quisque non lectus tincidunt, consequat sapien eu, rutrum nisl.","attachment":false,"time":"Tue Jan 05 2021 07:44:34 GMT+0530 (India Standard Time)","tag":"Jobs", "read": false, "important": false},
  {"id": 3, "from":1,"to":[3],"cc":[],"body":"Phasellus ullamcorper nisl a velit vulputate, in euismod felis sollicitudin.","attachment":false,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"", "read": false, "important": false},
  {"id": 4, "from":2,"to":[1],"cc":[],"body":"Integer lacinia, ex vitae interdum lobortis, felis nisi pretium tortor, et tincidunt dolor elit sed magna.","attachment":true,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"client", "read": true, "important": true},
  {"id": 5, "from":3,"to":[2],"cc":[],"body":"Integer lacinia, ex vitae interdum lobortis, felis nisi pretium tortor, et tincidunt dolor elit sed magna.","attachment":true,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"", "read": true, "important": true},
  {"id": 6, "from":2,"to":[3],"cc":[],"body":"Integer lacinia, ex vitae interdum lobortis, felis nisi pretium tortor, et tincidunt dolor elit sed magna.","attachment":true,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"", "read": true, "important": true},
  {"id": 7, "from":3,"to":[1],"cc":[2],"body":"Integer lacinia, ex vitae interdum lobortis, felis nisi pretium tortor, et tincidunt dolor elit sed magna.","attachment":true,"time":"Sat Jan 02 2021 18:02:02 GMT+0530 (India Standard Time)","tag":"", "read": true, "important": true}
]
