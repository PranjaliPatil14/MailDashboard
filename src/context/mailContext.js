import React, { useContext, useEffect, useReducer } from "react";
import * as PropTypes from "prop-types";
import useLocalStorage, { MAIL } from "../utilities/useLocalStorage";
import { UserContext } from "./userContext";
import {
  mailReducer,
  SET_LOGGED_IN_USER_MAIL,
  SET_VALUES,
} from "./mailReducer";

const initialMailContextValue = {
  mails: [],
  refresh: true,
  loggedInUserMails: [],
  dispatch: () => {},
};

export const MailContext = React.createContext(initialMailContextValue);

const MailContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(mailReducer, initialMailContextValue);
  const { value } = useLocalStorage(MAIL);
  const { loggedInUser } = useContext(UserContext);

  useEffect(() => {
    if (state.refresh) {
      dispatch({ type: SET_VALUES, data: value });
    }
  }, [value, state.refresh]);

  useEffect(() => {
    dispatch({ type: SET_LOGGED_IN_USER_MAIL, data: loggedInUser.id });
  }, [loggedInUser, value, state.mails]);

  useEffect(() => {
    localStorage.setItem(MAIL, JSON.stringify(state.mails));
  }, [state.mails]);

  return (
    <MailContext.Provider value={{ ...state, dispatch }}>
      {children}
    </MailContext.Provider>
  );
};

MailContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default MailContextProvider;
