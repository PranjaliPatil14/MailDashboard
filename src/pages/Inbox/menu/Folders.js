import React, { useContext } from "react";
import { Book, Circle, Inbox, Mail, Trash2 } from "react-feather";
import "./folder.css";
import { MailContext } from "../../../context/mailContext";
import Badge from "../../../components/badge/Badge";

const Folders = () => {
  const { loggedInUserMails } = useContext(MailContext);
  const folders = [
    {
      Icon: Inbox,
      title: "Inbox",
      badge: loggedInUserMails.filter(({ read }) => !read).length,
    },
    {
      Icon: Mail,
      title: "Sent Mails",
    },
    {
      Icon: Circle,
      title: "Important",
      badge: loggedInUserMails.filter(({ important }) => important).length,
    },
    {
      Icon: Book,
      title: "Drafts",
    },
    {
      Icon: Trash2,
      title: "Trash",
    },
  ];
  return (
    <div className="menu-folders">
      <p className="menu-folders-title">Folders</p>
      <ul className="folders">
        {folders.map(({ Icon, title, badge }) => (
          <li className="folder-item" key={title}>
            <Icon className="folder-item-icon" /> {title}
            {badge ? <Badge text={badge} /> : <></>}
          </li>
        ))}
      </ul>
    </div>
  );
};

export default Folders;
