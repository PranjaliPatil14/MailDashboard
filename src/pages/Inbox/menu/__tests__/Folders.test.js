import React from "react";
import { render } from "@testing-library/react";
import Folders from "../Folders";
import { MailContext } from "../../../../context/mailContext";
import mails from "../../../../mockData/mails.json";

describe("Folders", () => {
  const loggedInUserMails = mails;
  test("should render Folders", () => {
    const { asFragment } = render(
      <MailContext.Provider value={{ loggedInUserMails }}>
        <Folders />
      </MailContext.Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
