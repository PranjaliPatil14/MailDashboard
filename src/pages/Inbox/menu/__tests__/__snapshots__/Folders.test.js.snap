// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Folders should render Folders 1`] = `
<DocumentFragment>
  <div
    class="menu-folders"
  >
    <p
      class="menu-folders-title"
    >
      Folders
    </p>
    <ul
      class="folders"
    >
      <li
        class="folder-item"
      >
        <svg
          class="folder-item-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="22 12 16 12 14 15 10 15 8 12 2 12"
          />
          <path
            d="M5.45 5.11L2 12v6a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2v-6l-3.45-6.89A2 2 0 0 0 16.76 4H7.24a2 2 0 0 0-1.79 1.11z"
          />
        </svg>
         Inbox
        <span
          class="badge"
        >
          3
        </span>
      </li>
      <li
        class="folder-item"
      >
        <svg
          class="folder-item-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
          />
          <polyline
            points="22,6 12,13 2,6"
          />
        </svg>
         Sent Mails
      </li>
      <li
        class="folder-item"
      >
        <svg
          class="folder-item-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
        </svg>
         Important
        <span
          class="badge"
        >
          4
        </span>
      </li>
      <li
        class="folder-item"
      >
        <svg
          class="folder-item-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 19.5A2.5 2.5 0 0 1 6.5 17H20"
          />
          <path
            d="M6.5 2H20v20H6.5A2.5 2.5 0 0 1 4 19.5v-15A2.5 2.5 0 0 1 6.5 2z"
          />
        </svg>
         Drafts
      </li>
      <li
        class="folder-item"
      >
        <svg
          class="folder-item-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="3 6 5 6 21 6"
          />
          <path
            d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
          />
          <line
            x1="10"
            x2="10"
            y1="11"
            y2="17"
          />
          <line
            x1="14"
            x2="14"
            y1="11"
            y2="17"
          />
        </svg>
         Trash
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
